lib/ocamlorg_web/graphql.ml

```diff
@@ -0,0 +1,307 @@
+open Ocamlorg
+
+type package_dep =
+  { name : string
+  ; constraints : string option
+  }
+
+type packages_result =
+  { total_packages : int
+  ; packages : Package.t list
+  }
+
+let starts_with s1 s2 =
+  let len1 = String.length s1 in
+  if len1 > String.length s2 then
+    false
+  else
+    let s1 = String.lowercase_ascii s1 in
+    let s2 = String.lowercase_ascii s2 in
+    String.(equal (sub s2 0 (length s1)) s1)
+
+let is_package s1 s2 =
+  let len1 = String.length s1 in
+  if len1 <> String.length s2 then
+    false
+  else
+    let s1 = String.lowercase_ascii s1 in
+    let s2 = String.lowercase_ascii s2 in
+    String.(equal s2 s1)
+
+let get_packages_result total_packages offset limit filter packages =
+  match filter with
+  | None ->
+    let packages =
+      List.filteri (fun i _ -> offset <= i && i < offset + limit) packages
+    in
+    let packages_result = { total_packages; packages } in
+    packages_result
+  | Some filter ->
+    let packages =
+      List.filter
+        (fun package ->
+          starts_with filter (Package.Name.to_string (Package.name package)))
+        packages
+    in
+    let packages =
+      List.filteri (fun i _ -> offset <= i && i < offset + limit) packages
+    in
```

Comment:
> You could merge the two filterings into one by using `List.filteri` just once and making use of the second parameter of its filter function for the `starts_with` filtering. What do you think?

You mean like 
` let packages =
      List.filteri (fun i _ -> offset <= i && i < offset + limit) (List.filter (fun package ->
          starts_with filter (Package.Name.to_string (Package.name package)))
        packages)
    in
    let packages_result = { total_packages; packages } in
    packages_result`

Reply:
I mean something along the lines
```
let packages =
   List.filteri (fun i package ->
                        offset <= i && i < offset + limit &&
                        starts_with filter (Package.Name.to_string (Package.name package))
  ) packages
in

```

Reply:
This is cool but I think it will make that line to ambiguous to understand

Reply:
With "ambiguous", do you mean that someone reading the code could understand it in two different ways? What are those two ways?

An advantage of only filtering once would be a small improvement in performance: instead of going through the list twice, you only go through once. But to be fair, it probably wouldn't make a huge difference.

Reply:
@dinakajoy, have you seen this? :)

Reply:
Have you pushed that change? :) In the PR, there still seem to be the [two filters](https://github.com/ocaml/v3.ocaml.org-server/pull/60/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR42-R48): first the [starts_with filter](https://github.com/ocaml/v3.ocaml.org-server/pull/60/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR44) and then the [offset/limit filter](https://github.com/ocaml/v3.ocaml.org-server/pull/60/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR42).

Reply:
Hi @pitag-ha 
It was like this initially 
```List.filter
        (fun package ->
          starts_with filter (Package.Name.to_string (Package.name package)))
        packages
    in
    let packages =
      List.filteri (fun i _ -> offset <= i && i < offset + limit) packages
    in```
   
which was merged as this 
```List.filteri
        (fun i _ -> offset <= i && i < offset + limit)
        (List.filter
           (fun package ->
             starts_with filter (Package.Name.to_string (Package.name package)))
           packages)
    in```
    
    Maybe I misunderstood the merge you specified. Can you please clarify it a little bit more for me. Thank you

Reply:
Ok, that was a misunderstanding then. What I meant was only to filter only once as described in my comment https://github.com/ocaml/v3.ocaml.org-server/pull/60#discussion_r681833259. You're still filtering twice: once using filter and once using filteri.

Maybe have a quick look if you are able to implement this suggestion and if not I'd say to merge this PR as is and maybe to implement this suggestion in a separate (very short) PR.

Reply:
Perfect, thanks for implementing the change!